File: go/84. Largest Rectangle in Histogram.go
func largestRectangleArea(heights []int) int {
    stack := NewStack()
	answer := 0
	i := 0
	for ; i < len(heights); i++ {
		for !stack.Empty() && heights[i] <= heights[stack.Top()] {
			height := heights[stack.Pop()]
			if stack.Empty() {
				answer = max(answer, i * height)
			} else {
				answer = max(answer, height * (i - stack.Top() - 1))
			}
		}
		stack.Push(i)
	}
	for !stack.Empty() {
		height := heights[stack.Pop()]
		if stack.Empty() {
			answer = max(answer, i * height)
		} else {
			answer = max(answer, height * (i - stack.Top() - 1))
		}
	}
	return answer
}

type Stack struct {
	data []int
}

func NewStack() *Stack {
	return &Stack{
		data: []int{},
	}
}

func (s *Stack) Push(x int) {
	s.data = append(s.data, x)
}

func (s *Stack) Pop() int {
	if len(s.data) == 0 {
		return -1
	}
	x := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	return x
}

func (s *Stack) Top() int {
	if len(s.data) == 0 {
		return -1
	}
	return s.data[len(s.data)-1]
}

func (s *Stack) Empty() bool {
	return len(s.data) == 0
}

File: java/84. Largest Rectangle in Histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        int answer = 0;
        int i = 0;
        for (; i < heights.length; i++) {
            while (!stack.isEmpty() && heights[i] <= heights[stack.peekLast()]) {
                int height = heights[stack.pollLast()];
                if (stack.isEmpty()) answer = Math.max(answer, i * height);
                else answer = Math.max(answer, height*(i-stack.peekLast()-1));
            }
            stack.addLast(i);
        }
        while (!stack.isEmpty()) {
            int height = heights[stack.pollLast()];
            if (stack.isEmpty()) answer = Math.max(answer, i * height);
            else answer = Math.max(answer, height * (i-stack.peekLast()-1));
        }
        return answer;
    }
}